gspread/exceptions.py

```diff
@@ -40,7 +41,19 @@ class APIError(GSpreadException):
     such as when we attempt to retrieve things that don't exist."""
 
     def __init__(self, response: Response):
-        error = dict(response.json()["error"])
+        try:
+            error = response.json()["error"]
+        except JSONDecodeError:
+            # in case we failed to parse the error from the API
+            # build an empty error object to notify the caller
+            # and keep the exception raise flow running
+
+            error = {
+                "code": -1,
```

Comment:
Hey @lavigne958 I'm wondering if the error code here should instead be `response.status_code`. I noticed this because the BackOffHTTPClient will not retry to send the request because -1 is not an expected error code [here](https://github.com/burnash/gspread/blob/master/gspread/http_client.py#L573).

Reply:
Hi that's a good point. 

Unfortunately we can't simply set the return code as we failed to parse the returned error.

This could be 2 reasons: 1. The response is incomplete/connection lost etc. 2. The response is in html or some other format we don't handle nor expect but the response is valid !

I will check with alifee what is best here, I have 2 ideas:
1. Instead of returning a mock http response, we raise a new Gspread issue and the user is in charge of retrying or not. We don't take a decision here.

2. We add a new setting for the http client to retry on -1 errors. So the user can enable that setting and the http client will retry on -1 error code too.

@alifee what you do think ?

Reply:
For what it's worth, I imagine there would still be HTTP codes coming back with the scenarios you mentioned, e.g. 408 (timeout), 502 (server error, with html response).